enhancements/update/eus-upgrades-mvp.md

```diff
@@ -0,0 +1,400 @@
+  and operator managed components (Kubernetes API, OpenShift API, etc)
+  - Respecting version skew policies of OLM managed Operators
+  - Ensuring APIs removed by components in the release payload are no longer in
+  use
+- We will ensure that CVO managed operators upgrade as quickly as is safe to do so
+- We will reduce workload disruption associated with pod rescheduling during rolling
+reboots
+
+
+### Non-Goals
+
+- This enhancement does not attempt to remove *any* steps along the serial
+  upgrade path from 4.6 to 4.7 to 4.8 to 4.9 to 4.10 *including* reboots. A
+  follow-up enhancement will address those concerns.
+
+## Proposal
+
+Before we dive into user stories it's useful that you're familiar with CVO and
+Console features which are relevant to how upgrades and channel options are
+presented to the admin. The CVO and Console as of 4.6 are now aware of other
+channels containing the cluster's current version. This means that if 4.6.50 is
+a non EUS version and 4.6.51 is an EUS version both the CVO and Console can
+display a different set of channel options. This means that 4.6.50 would see
+candidate-4.6, fast-4.6, stable-4.6, eus-4.6, candidate-4.7, fast-4.7, and
+stable-4.7 channels.
+
+Where as 4.6.51 would see only eus-4.6 and eus-4.10 assuming we have enabled EUS
+4.6 to 4.10 upgrades. After switching to eus-4.10 the cluster would offer
+upgrades to 4.7 versions present in the eus-4.10 channel, when that completes
+the cluster would offer 4.8 versions present in the eus-4.10 channel, repeat
+until the cluster has completed its upgrade to 4.10.
+
+### User Stories
+
+#### APIServer - Notify cluster admins of API use which will be removed between 4.7 and 4.10
+
+We have pre-existing mechanisms to alert admins to their use of APIs which are
+slated for removal in a future release. Currently those mechanisms do not look
+forward far enough to span 4.6 to 4.10, we will need to add the ability to
+backport that API metadata to 4.6 once we have a firm understanding of all
+removals between 4.6 and 4.10.
+
+In addition to Info level alerts core API providers will also need to set
+Upgradeable=False whenever the next minor version removes an API which is in
+use.
+
+This allows us to *inform* the admin for removals that are more than one minor
+version away and *block* upgrades for removals which are imminent.
+
+### MCO - Enforce OpenShift's defined host component version skew policies
+
+The MCO, will set Upgradeable=False whenever any MachineConfigPool has one more
+more nodes present which fall outside of a defined list of constraints. For
+instance, if OpenShift has a defined Kubelet Version Skew of N-1, the node
+constraints enforced by the MCO defined in OCP 4.7 could be as follows:
+
+```yaml
+node.status.nodeInfo.kubeletVersion:
+- v1.20
+- v1.19
+```
+
+If the policy were to change to allow for a version skew of N-2, v1.18 would be
+added to the list of acceptable matches. The MCO is not responsible for defining
```

Comment:
How will these constrains be defined, and how does the MCO retroactively apply them to prevent the upgrade in the first place?

It sounds like the 4.6 MCO would first have to understand what the constraints are to prevent the upgrade to 4.7, despite the restrictions defined in the 4.7 MCO?

Reply:
Actually on second thought, there is no real need for the MCO to expand on its current block on pools, since a previous upgrade not finishing would block further upgrades anyways. This means we would never have a kubelet version skew > 1, making this check pointless (unless you wanted to initiate a direct update from 4.6->4.8, which as I understand, shouldn't happen regardless).

Reply:
The specific implementation needs to be hashed out, but it would be expected to be defined in the release payload of the version of the MCO which is running, a configmap most likely.